### PR TITLE
Merge in NBitcoin changes up to v4.0.0.54

### PR DESCRIPTION
**Pull request by mikedennis**
These are mostly around transaction builder:

https://github.com/MetacoSA/NBitcoin/commit/40520597e4d48f3e4789a467f4acf0eda1d7bbca
https://github.com/MetacoSA/NBitcoin/commit/b08eb8b378b0f6c3ede97f06eb947043c1641879
https://github.com/MetacoSA/NBitcoin/commit/fc5fc5b6fa389e515f164fbfe98b844053fc6dbe
https://github.com/MetacoSA/NBitcoin/commit/0d7f89d16feb2d62ef4ed42ec9a7e4db5e4d6cd0
https://github.com/MetacoSA/NBitcoin/commit/950c2bd124a915d48e493f45d766af6239b7a768
https://github.com/MetacoSA/NBitcoin/commit/c105ac67d2e40d5a57cd682c85108799bfc81e6c

### CODE REVIEW

**Comment by mikedennis** [comment]
@bokobza Hmmm this change seems to break a wallet test. This assertion is now failing, there are now 4 inputs instead of 3.

https://github.com/stratisproject/StratisBitcoinFullNode/blob/master/src/Stratis.Bitcoin.Features.Wallet.Tests/WalletTransactionHandlerTest.cs#L275

Not sure I can explain why there are 4 based upon these changes. Do you know? Investigating further...

**Comment by dangershony** [comment]
We have to be very careful as there are changes to the coin selection algorithm.

**Comment by mikedennis** [comment]
I see there is another transaction builder fix since I created this PR. https://github.com/MetacoSA/NBitcoin/pull/407 Seems like more to do with using the builder incorrectly but let me know if you think it's worthwhile to add it to this PR.

**Comment by dangershony** [comment]
You you can bring it over why not, looks like a small improvement.

**Comment by dangershony** [comment]
Mike can you confirm all tests pass? most importantly the ones that use the builder

**Comment by mikedennis** [comment]
There are two integration tests that are failing that I need to investigate:
```
Test Name:	A_nulldata_transaction_is_sent_to_the_network
Test FullName:	Stratis.Bitcoin.IntegrationTests.Transactions.TransactionWithNullDataSpecification.A_nulldata_transaction_is_sent_to_the_network
Test Source:	D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\Transactions\TransactionWithNullDataSpecification.cs : line 11
Test Outcome:	Failed
Test Duration:	0:00:33.588

Result StackTrace:	
at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Stratis.Bitcoin.IntegrationTests.Transactions.TransactionWithNullDataSpecification.the_transaction_should_get_confirmed() in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\Transactions\TransactionWithNullDataSteps.cs:line 120
   at Stratis.Bitcoin.IntegrationTests.TestFramework.BddSpecification.RunStep(Action step, String stepType) in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\TestFramework\BddSpecification.cs:line 93
   at Stratis.Bitcoin.IntegrationTests.TestFramework.BddSpecification.Then(Action step) in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\TestFramework\BddSpecification.cs:line 72
   at Stratis.Bitcoin.IntegrationTests.Transactions.TransactionWithNullDataSpecification.A_nulldata_transaction_is_sent_to_the_network() in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\Transactions\TransactionWithNullDataSpecification.cs:line 21
Result Message:	System.InvalidOperationException : Sequence contains no elements
Result StandardOutput:	
(5:37:55 AM) Given two proof of work nodes
(5:37:59 AM) And a sending and a receiving wallet
(5:38:07 AM) And some funds in the sending wallet
(5:38:16 AM) And no fund in the receiving wallet
(5:38:16 AM) And the wallets are in sync
(5:38:20 AM) And a nulldata transaction
(5:38:22 AM) When the transaction is broadcasted
(5:38:22 AM) And the block is mined
(5:38:25 AM) Then the transaction should get confirmed
(5:38:29 AM) [End of test - 33.6104981 seconds.]
```

and

```
Test Name:	A_transaction_can_be_retrieved_by_its_identifier_and_a_wrong_identifier_returns_null
Test FullName:	Stratis.Bitcoin.IntegrationTests.BlockStore.RetrieveFromBlockStoreSpecification.A_transaction_can_be_retrieved_by_its_identifier_and_a_wrong_identifier_returns_null
Test Source:	D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\BlockStore\RetrieveFromBlockStoreSpecification.cs : line 27
Test Outcome:	Failed
Test Duration:	0:00:40.903

Result StackTrace:	
at FluentAssertions.Execution.XUnit2TestFramework.Throw(String message) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions\Execution\XUnit2TestFramework.cs:line 32
   at FluentAssertions.Execution.AssertionScope.FailWith(String message, Object[] args) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions\Execution\AssertionScope.cs:line 224
   at FluentAssertions.Primitives.ReferenceTypeAssertions`2.NotBeNull(String because, Object[] becauseArgs) in C:\projects\fluentassertions-vf06b\Src\FluentAssertions\Primitives\ReferenceTypeAssertions.cs:line 65
   at Stratis.Bitcoin.IntegrationTests.BlockStore.RetrieveFromBlockStoreSpecification.the_real_transaction_should_be_retrieved() in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\BlockStore\RetrieveFromBlockStoreSteps.cs:line 186
   at Stratis.Bitcoin.IntegrationTests.TestFramework.BddSpecification.RunStep(Action step, String stepType) in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\TestFramework\BddSpecification.cs:line 93
   at Stratis.Bitcoin.IntegrationTests.TestFramework.BddSpecification.Then(Action step) in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\TestFramework\BddSpecification.cs:line 72
   at Stratis.Bitcoin.IntegrationTests.BlockStore.RetrieveFromBlockStoreSpecification.A_transaction_can_be_retrieved_by_its_identifier_and_a_wrong_identifier_returns_null() in D:\Devel\source\stratis-fork\StratisBitcoinFullNode\src\Stratis.Bitcoin.IntegrationTests\BlockStore\RetrieveFromBlockStoreSpecification.cs:line 41
Result Message:	Expected this.retrievedTransaction not to be <null>.
Result StandardOutput:	
(5:38:36 AM) Given a pow node running
(5:38:39 AM) And a pow node to transact with
(5:38:45 AM) And a miner validating blocks
(5:38:51 AM) And some blocks creating reward
(5:39:00 AM) And the nodes are synced
(5:39:06 AM) And a real transaction
(5:39:09 AM) And a wrong transaction id
(5:39:09 AM) And the block with the transaction is mined
(5:39:13 AM) When trying to retrieve the transactions by Id from the blockstore
(5:39:13 AM) And trying to retrieve the block containing the transactions from the blockstore
(5:39:13 AM) Then the real transaction should be retrieved
(5:39:17 AM) [End of test - 40.9205115 seconds.]
```

**Comment by dangershony** [comment]
Hi Mike this will have to go in after the static flags change and likely you will have to merge some code

**Comment by mikedennis** [comment]
Integration test failure is related to this new feature in DefaultCoinSelector in TransactionBuilder:
```
      /// <summary>
        /// Select all coins belonging to same scriptPubKey together to protect privacy. (Default: true)
        /// </summary>
        public bool GroupByScriptPubKey
        {
            get; set;
        } = true;
```
If I set it to false the tests pass - no exception is thrown.

If it's set to true then this exception is thrown: ```
Stratis.Bitcoin.Features.Consensus.ConsensusErrorException: 'tried to spend coinbase before maturity' ```

I'm having trouble figuring out whether it's an issue with the test or a bug with the coin selection algorithm. I suspect possibly the test needs a modification as all the other tests pass except the two mentioned above.

**Comment by mikedennis** [comment]
This is my theory of what is happening:
This test is a bit of an artificial scenario in that all the coins are created with the same scriptpubkey. Therefore the new privacy features grabs all the coins including the coinbase. The consensus validator then throws an exception saying you are trying to spend coinbase before maturity but it's not actually spending it, it's just in the transaction for obfuscation. I'm not sure if this could happen in reality or whether this is just an artifact of this test?
I do notice in many of NBitcoin's tests he is turning this off in the tx builder like so ``` ((DefaultCoinSelector)txBuilder.CoinSelector).GroupByScriptPubKey = false; ```
More difficult for us in these tests, because we are using the WalletTransactionHandler to create the transaction builder so it's isolated from the tests - I think we would have to pass in a flag in the TransactionBuildContext.

Looking for your thoughts @dangershony and @bokobza on whether I'm on the right track here and whether it's a test issue or a bug in the transaction builder we need to discuss with Nicolas.

**Comment by mikedennis** [comment]
I've fixed the broken tests. The tests were using a tx confirmation number less than coinbase maturity which was why the unmature txs were being included. It worked before because the coin selection order was different so it didn't ever grab the coinbase txs. All integration tests should now pass.

I'm not sure whether this is something the wallet should be checking in this method or whether it's the callers responsibility to make sure confirmations are greater than maturity?

Wallet.cs
```
        /// <summary>
        /// Lists all spendable transactions in the current account.
        /// </summary>
        /// <param name="currentChainHeight">The current height of the chain. Used for calculating the number of confirmations a transaction has.</param>
        /// <param name="confirmations">The minimum number of confirmations required for transactions to be considered.</param>
        /// <returns>A collection of spendable outputs that belong to the given account.</returns>
        public IEnumerable<UnspentOutputReference> GetSpendableTransactions(int currentChainHeight, int confirmations = 0)
```

**Comment by dangershony** [comment]
ok please resolve conflicts and we try to push it soon (its open for a while now)

File: src/NBitcoin.Tests/transaction_tests.cs [hunk 1]
```diff
@@ -200,13 +200,25 @@ public void CanSelectCoin()
                 CreateCoin("3.0"),
                 CreateCoin("3.0")
             }, Money.Parse("10.0")));
+
```
Comment (noescape00) [inline src/NBitcoin.Tests/transaction_tests.cs]: redundant line

File: src/NBitcoin/RPC/RPCException.cs [hunk 2]
```diff
@@ -3,47 +3,170 @@
...
-        RPC_CLIENT_NOT_CONNECTED = -9, // Bitcoin is not connected
-        RPC_CLIENT_IN_INITIAL_DOWNLOAD = -10, // Still downloading initial blocks
-        RPC_CLIENT_NODE_ALREADY_ADDED = -23, // Node is already added
-        RPC_CLIENT_NODE_NOT_ADDED = -24, // Node has not been added before
+        /// <summary>
+        /// std::exception thrown in command handling
+        /// </summary>
+        RPC_MISC_ERROR = -1,
+        /// <summary>
+        /// Server is in safe mode, and command is not allowed in safe mode
```
Comment (noescape00) [inline src/NBitcoin/RPC/RPCException.cs]: Missing dots in all the summaries

File: src/NBitcoin/RPC/RPCException.cs [hunk 3]
```diff
@@ -3,47 +3,170 @@
...
-        RPC_TRANSACTION_ALREADY_IN_CHAIN = -27, // Transaction already in chain
+        //! General application defined errors
 
-        // P2P client errors
-        RPC_CLIENT_NOT_CONNECTED = -9, // Bitcoin is not connected
-        RPC_CLIENT_IN_INITIAL_DOWNLOAD = -10, // Still downloading initial blocks
-        RPC_CLIENT_NODE_ALREADY_ADDED = -23, // Node is already added
-        RPC_CLIENT_NODE_NOT_ADDED = -24, // Node has not been added before
+        /// <summary>
+        /// std::exception thrown in command handling
```
Comment (noescape00) [inline src/NBitcoin/RPC/RPCException.cs]: ` std::exception` comment probably came from bitcoin core.

File: src/NBitcoin/TransactionBuilder.cs [hunk 4]
```diff
@@ -578,13 +594,32 @@ public Network Network
...
         {
             get;
             set;
         }
 
+        /// <summary>
+        /// If true, the TransactionBuilder will not select coins whose fee to spend is higher than its value. (Default: true)
+        /// The cost of spending a coin is based on the <see cref="FilterUneconomicalCoinsRate"/>.
+        /// </summary>
+        public bool FilterUneconomicalCoins
```
Comment (noescape00) [inline src/NBitcoin/TransactionBuilder.cs]: please convert this guy to be in one line 🙂

**Comment by noescape00** [comment]
Also please fix merge conflicts.

**Pull request merged**